### PR TITLE
Overhaul revisions and annotations

### PR DESCRIPTION
**Pull request by neongreen**
A paste has annotations and revisions (the original doesn't count as a revision – therefore, `getRevisions` has been changed). Everywhere the original version of a paste is referred to as `original`, and the latest version as `latest` (the new function `getLatestVersion` gets the latest version of a paste). Pastes are referred to by the ID of the original paste.

This pull request does the following:

  * A new type, `PasteContext`, is introduced. `PasteContext` is basically everything that user sees on a paste page – it consists of the paste itself + its annotations + its revisions + hints for all that. This type makes it easier to handle situations when we need access both to the original paste and the latest paste.

  * We resolve the latest versions of annotations (because annotations can be revised too) and show them instead of original pastes.

  * When editing/annotating an annotation, we redirect to the *parent* paste of the annotation (and don't forget to invalidate the cache for it).

  * When editing a paste, we fill in the title (I don't know why it wasn't filled in previously, but that was inconvenient for small edits).

  * Both the original author of a paste and the author of the latest revision are displayed on the “Browse” page.

### CODE REVIEW

**Comment by neongreen** [comment]
Oh, and also now editing annotations actually *works*, tho it's more of a side effect of all other changes :)

File: src/Hpaste/Controller/Paste.hs [hunk 1]
```diff
@@ -48,34 +51,65 @@ handle revision = do
 	      	       	    else getPasteById (pid)
         case paste of
```
Comment (Gurkenglas) [inline src/Hpaste/Controller/Paste.hs]: paste `for` \pcOriginal -> model $ do
  ...
  return $ page PastePage
    ...

File: src/Hpaste/Controller/Paste.hs [hunk 2]
```diff
@@ -48,34 +51,65 @@ handle revision = do
...
+  -> Maybe (Paste, Paste)
+  -> HPCtrl Html
 pasteForm channels languages defChan annotatePaste editPaste = do
   params <- getParams
   submittedPrivate <- isJust <$> getParam "private"
   submittedPublic <- isJust <$> getParam "public"
-  revisions <- maybe (return []) (model . getRevisions) (fmap pasteId (annotatePaste <|> editPaste))
+  let mbOriginal = fst <$> (annotatePaste <|> editPaste)
+      mbLatest   = snd <$> (annotatePaste <|> editPaste)
+  let parentPasteId = case mbOriginal of
```
Comment (Gurkenglas) [inline src/Hpaste/Controller/Paste.hs]:
```
parentPasteId = mbOriginal >>= \p -> case pasteType p of
  AnnotationOf x -> Just x
  _ -> Nothing
```

(Or even `mbOriginal >>= preview AnnotationOf`, but some people don't like lens.)

File: src/Hpaste/Controller/Paste.hs [hunk 3]
```diff
@@ -101,9 +135,10 @@ pasteForm channels languages defChan annotatePaste editPaste = do
       	 then goSpamBlocked
 	 else do
 	    resetCache Key.Home
-	    maybe (return ()) (resetCache . Key.Paste) $ pasteSubmitId paste
+	    mapM_ (resetCache . Key.Paste) $
+              nub $ catMaybes [pasteSubmitId paste, parentPasteId]
 	    pid <- model $ createPaste languages channels paste spamrating submittedPublic
-	    maybe (return ()) redirectToPaste pid
+	    maybe (return ()) redirectToPaste (parentPasteId <|> pid)
```
Comment (Gurkenglas) [inline src/Hpaste/Controller/Paste.hs]: mapM_ redirectToPaste (parentPasteId <|> pid)

File: src/Hpaste/Controller/Raw.hs [hunk 4]
```diff
@@ -26,6 +26,10 @@ handle = do
   case pid of
```
Comment (Gurkenglas) [inline src/Hpaste/Controller/Raw.hs]:
```
handle = fromMaybeT goHome $ do
  pid <- MaybeT (getParam "id") >>= liftMaybe . readMay . toString
  paste <- MaybeT $ model $ getLatestVersionById (PasteId pid)
  lift $ modifyResponse $ setContentType "text/plain; charset=UTF-8"
  lift $ outputText . fromStrict . pastePaste $ paste
```

Of course, fromMaybeT and liftMaybe don't exist yet.

File: src/Hpaste/View/Paste.hs [hunk 5]
```diff
@@ -184,48 +180,63 @@ getPasteId PasteFormlet{..} =
...
-	    	    | otherwise      = map pasteAuthor revisions
-        htmlCommasAnd $ flip map (nub authors) $ \author ->
-	  linkAuthor author
-      detail "Language" $ showLanguage langs (pasteLanguage paste)
-      detail "Channel" $ showChannel (Just paste) chans (pasteChannel paste)
-      detail "Created" $ showDateTime (pasteDate paste)
-      unless (length revisions < 2) $ detail "Revisions" $ do
+        linkToParent latest
+      let authors = nub (map pasteAuthor (original : pcRevisions paste))
+      detail (if length authors > 1 then "Authors" else "Author") $ do
```
Comment (Gurkenglas) [inline src/Hpaste/View/Paste.hs]: Why the `do`

File: src/Hpaste/View/Paste.hs [hunk 6]
```diff
@@ -275,57 +278,50 @@ pasteNav pastes paste =
...
-	        when (pasteType paste /= AnnotationOf prevId) $ do
-                  " / "
-                  href ("/diff/" ++ show prevId ++ "/" ++ show pid)
-                       ("prev" :: Text)
-            case listToMaybe pastes of
-              Nothing -> return ()
-              Just{} -> " - "
-
+    where pid = pasteId (pcOriginal paste)
+          diffLink = case mbAnnotationInfo of
```
Comment (Gurkenglas) [inline src/Hpaste/View/Paste.hs]: diffLink = mbAnnotationInfo `for_` \(prevAnnId, parentPasteId) -> do ...

File: src/Hpaste/View/Paste.hs (hunk 5 above)
Comment (neongreen) [inline src/Hpaste/View/Paste.hs]: To be able to write `let` without `in`.

Comment (neongreen) [inline src/Hpaste/View/Paste.hs]: No, wait, nevermind, was looking at the wrong part of the diff. Right, this `do` isn't needed here.

File: src/Hpaste/View/Paste.hs (hunk 6 above)
Comment (neongreen) [inline src/Hpaste/View/Paste.hs]: I think the `case` is more readable here (at least for me).

**Comment by neongreen** [comment]
These commits are somewhat messy, so I've started breaking them into more fine-grained commits so that it'd be easier to see what exactly I've changed. I'll push an updated vesion in a few days.

**Pull request closed without merging**